src/backend/distributed/executor/local_executor.c

```diff
@@ -0,0 +1,579 @@
+	ExecutorRun(queryDesc, scanDirection, 0L, true);
+
+	/*
+	 * We'll set the executorState->es_processed later, for now only remember
+	 * the count.
+	 */
+	if (taskPlan->commandType != CMD_SELECT)
+	{
+		totalRowsProcessed = queryDesc->estate->es_processed;
+	}
+
+	ExecutorFinish(queryDesc);
+	ExecutorEnd(queryDesc);
+
+	FreeQueryDesc(queryDesc);
+
+	return totalRowsProcessed;
+}
+
+
+/*
+ *  ShouldExecuteTasksLocally gets a distributed plan and returns true if the
+ *  plan should be executed locally. This function does not guarantee that any
+ *  task have to be executed locally.
+ */
+bool
+ShouldExecuteTasksLocally(DistributedPlan *distributedPlan)
+{
+	List *taskList = NIL;
+	bool singleTask = false;
+
+	if (!EnableLocalExecution)
+	{
+		return false;
+	}
+
+	if (GetLocalGroupId() == 0)
+	{
+		/* currently local execution is only supported on metadata workers */
+		return false;
+	}
+
+	if (LocalExecutionHappened)
+	{
+		/*
+		 * For various reasons, including the transaction visibility
+		 * rules (e.g., read-your-own-writes), we have to use local
+		 * execution again if it has already happened within this
+		 * transaction block.
+		 *
+		 * We might error out later in the execution if it is not suitable
+		 * to execute the tasks locally.
+		 */
+		Assert(IsMultiStatementTransaction() || InCoordinatedTransaction());
+
+		/*
+		 * TODO: A future improvement could be to keep track of which placements
+		 * have been locally executed. At this point, only use local execution for
+		 * those placements. That'd help to benefit more from parallelism.
+		 */
+
+		return true;
+	}
+
+	taskList = distributedPlan->workerJob->taskList;
+	singleTask = (list_length(taskList) == 1);
+	if (singleTask && TaskAccessLocalNode((Task *) linitial(taskList)))
+	{
+		/*
+		 * This is the valuable time to use the local execution. We are likely
+		 * to avoid any network round-trips by simply executing the command within
+		 * this session.
+		 *
+		 * We cannot avoid network round trips if the task is not a read only
+		 * task and accesses multiple placements. For example, modifications to
+		 * distributed tables (with replication factor == 1) would avoid network
+		 * round-trips. However, modifications to reference tables still needs to
+		 * go to over the network to do the modification on the other placements.
+		 * Still, we'll be avoding the network round trip for this node.
+		 *
+		 * Note that we shouldn't use local execution if any distributed execution
+		 * has happened because that'd break transaction visibility rules and
+		 * many other things.
+		 */
+		return !AnyConnectionAccessedPlacements();
+	}
+
+	if (!singleTask)
+	{
+		/*
+		 * For multi-task executions, switching to local execution would likely to
+		 * perform poorly, because we'd lose the parallelizm. Note that the local
+		 * execution is happening one task at a time (e.g., similar to sequential
+		 * distributed execution).
+		 */
+		Assert(!LocalExecutionHappened);
```

Comment:
Given the `if (LocalExecutionHappened) return true;` above, this assert may be a little redundant.

Reply:
I kept it just in case someone tries to re-order the function in the future. I can remove if this looks weird.